src/keystore/persistence/EncryptedPersistence.ts

```diff
@@ -0,0 +1,104 @@
+import eccrypto, { Ecies } from 'eccrypto'
+
+const IV_LENGTH = 16
+const EPHEMERAL_PUBLIC_KEY_LENGTH = 65
+const MAC_LENGTH = 32
+const AES_BLOCK_SIZE = 16
+const MIN_DATA_LENGTH =
+  IV_LENGTH + EPHEMERAL_PUBLIC_KEY_LENGTH + MAC_LENGTH + AES_BLOCK_SIZE
+
+const assertEciesLengths = (ecies: Ecies): void => {
+  if (ecies.iv.length !== IV_LENGTH) {
+    throw new Error('Invalid iv length')
+  }
+  if (ecies.ephemPublicKey.length !== EPHEMERAL_PUBLIC_KEY_LENGTH) {
+    throw new Error('Invalid ephemPublicKey length')
+  }
+  if (
+    ecies.ciphertext.length < 1 ||
+    ecies.ciphertext.length % AES_BLOCK_SIZE !== 0
+  ) {
+    throw new Error('Invalid ciphertext length')
+  }
+  if (ecies.mac.length !== MAC_LENGTH) {
+    throw new Error('Invalid mac length')
+  }
+}
+
+const serializeEcies = (ecies: Ecies): Uint8Array => {
+  assertEciesLengths(ecies)
+  const { iv, ephemPublicKey, ciphertext, mac } = ecies
+  const result = new Uint8Array(
+    iv.length + ephemPublicKey.length + ciphertext.length + mac.length
+  )
+  result.set(iv, 0)
+  result.set(ephemPublicKey, IV_LENGTH)
+  result.set(ciphertext, IV_LENGTH + EPHEMERAL_PUBLIC_KEY_LENGTH)
+  result.set(mac, IV_LENGTH + EPHEMERAL_PUBLIC_KEY_LENGTH + ciphertext.length)
+  return result
+}
+
+const deserializeEcies = (data: Uint8Array): Ecies => {
+  if (
+    data.length < MIN_DATA_LENGTH ||
+    (data.length - MIN_DATA_LENGTH) % AES_BLOCK_SIZE !== 0
+  ) {
+    throw new Error('Invalid data length')
+  }
+  const iv = data.slice(0, IV_LENGTH)
+  const ephemPublicKey = data.slice(
+    IV_LENGTH,
+    IV_LENGTH + EPHEMERAL_PUBLIC_KEY_LENGTH
+  )
+  const ciphertext = data.slice(
+    IV_LENGTH + EPHEMERAL_PUBLIC_KEY_LENGTH,
+    data.length - MAC_LENGTH
+  )
+  const mac = data.slice(data.length - MAC_LENGTH, data.length)
+  return {
+    iv: Buffer.from(iv),
+    ephemPublicKey: Buffer.from(ephemPublicKey),
+    ciphertext: Buffer.from(ciphertext),
+    mac: Buffer.from(mac),
+  }
+}
+
+/**
+ * EncryptedPersistence is a Persistence implementation that uses ECIES to encrypt all values
+ */
+export default class EncryptedPersistence implements Persistence {
+  private persistence: Persistence
+  private privateKey: Buffer
+  private publicKey: Buffer
+
+  constructor(persistence: Persistence, privateKey: Uint8Array) {
+    this.persistence = persistence
+    this.privateKey = Buffer.from(privateKey)
+    this.publicKey = eccrypto.getPublic(this.privateKey)
+  }
+
+  async getItem(key: string): Promise<Uint8Array | null> {
+    const encrypted = await this.persistence.getItem(key)
+    if (encrypted) {
+      return this.decrypt(encrypted)
+    }
+    return null
+  }
+
+  async setItem(key: string, value: Uint8Array): Promise<void> {
+    const encrypted = await this.encrypt(value)
+    await this.persistence.setItem(key, encrypted)
+  }
+
+  private async encrypt(value: Uint8Array): Promise<Uint8Array> {
+    const ecies = await eccrypto.encrypt(this.publicKey, Buffer.from(value))
```

Comment:
`LocalStorage` persistence (and the potential future `FilesystemPersistence`) both have the possibility of malicious entities deleting the store or inserting bad data. For example, another script on the page can write anything it wants to `LocalStorage`.

I'm not sure I understand what problem authentication would solve. If the malicious entity encrypted their data with another key or wrote garbage to the storage, decryption would fail and the process would bail out. That feels like the correct behaviour.

Reply:
You're totally right. Someone could use a valid private key and encrypt a message for the user's public key that is decryptable. 

I still don't see where the attack is. Making up topics and putting them into the conversation store feels more like mischief than an attack, with the way the Keystore works.

Reply:
At least for the current use-case I care about unauthenticated reads and care less about unauthenticated writes. But I can't speak to other use-cases for the persistence. 

Do we have a clear idea of how we would protect against both if we wanted to future proof the solution?

Reply:
That wouldn't be too hard. Given that this code isn't used yet, it's safe and easy for me to merge and then continue to improve the implementation.

Reply:
From a threat perspective there are a few cases to consider

Fake Entry generation: An attacker generates malicious entries. 
- Would be immediately detectable since the attacker does not have access to the the other valid entries
- Harm is injecting a topic/topic key into the inbox. Since topic keys are not updated, and attacker could DOS a conversation by blocking the correct key form being entered into the store. 

Persistence Entry Replay: An attacker deletes the current entry with an previous blob. 
- End result is a few missing entries which I imagine would be re-added once a KeyStore recieves a message it doesn't have a key for 

Persistence Entry Deletion: Attacker wipes localstorage
- Keys are rebuild
- No harm

Adding a signature would help with 1 but the placement and replay protection is something we'd want to consider before implementing

Reply:
OK. I'm going to merge as-is and make a ticket to add a signature. Will get that in before v8 is released.

Looking at this some more, I think there is an attack vector where an attacker could forge an invite to a conversation purporting to be with user X, but with weakened encryption that the attacker can read. They would then be able to read one side of the conversation and write to it.

Reply:
Sounds good. 

> and write to it
I'd be be interested to see the analysis on this. 
Messages are signed by the identity key of the sender, so an attacker would need access to the Identity key if they are attempting to spoof a message. 

If a spoofed message can be injected into an inbox then that should be addressed

Reply:
Ya. You're right. They could write to the topic, but the signatures would fail to verify and the message would be discarded. Could read though.